rust/index/src/fulltext/types.rs

```diff
@@ -27,318 +29,201 @@ impl ChromaError for FullTextIndexError {
         &self,
-        tokens: &[Token],
+        mutations: M,
     ) -> Result<(), FullTextIndexError> {
-        // (Scoped to limit the lifetime of the lock)
-        {
-            let mut uncommitted_frequencies = self.uncommitted_frequencies.lock().await;
-            for token in tokens {
-                if uncommitted_frequencies.contains_key(&token.text) {
-                    continue;
+        let mut token_instances = vec![];
+
+        for mutation in mutations {
+            match mutation {
+                DocumentMutation::Create {
+                    offset_id,
+                    new_document,
+                } => {
+                    self.tokenizer
+                        .clone()
+                        .token_stream(new_document)
+                        .process(&mut |token| {
+                            token_instances.push(TokenInstance::encode(
+                                token.text.as_str(),
+                                offset_id,
+                                Some(token.offset_from as u32),
+                            ));
+                        });
                 }
 
-                let frequency = match &self.full_text_index_reader {
-                    // Readers are uninitialized until the first compaction finishes
-                    // so there is a case when this is none hence not an error.
-                    None => 0,
-                    Some(reader) => {
-                        (reader.get_frequencies_for_token(token.text.as_str()).await).unwrap_or(0)
-                    }
-                };
-                uncommitted_frequencies
-                    .insert(token.text.clone(), (frequency as i32, frequency as i32));
-            }
-        }
-
-        let mut uncommitted_postings = self.uncommitted_postings.lock().await;
-        for token in tokens {
-            if uncommitted_postings
-                .positional_postings
-                .contains_key(&token.text)
-            {
-                continue;
-            }
+                DocumentMutation::Update {
+                    offset_id,
+                    old_document,
+                    new_document,
+                } => {
+                    // Remove old version
+                    let mut trigrams_to_delete = HashSet::new(); // (need to filter out duplicates, each trigram may appear multiple times in a document)
+                    self.tokenizer
+                        .clone()
+                        .token_stream(old_document)
+                        .process(&mut |token| {
+                            trigrams_to_delete.insert(TokenInstance::encode(
+                                token.text.as_str(),
+                                offset_id,
+                                None,
+                            ));
+                        });
+
+                    // Add doc
+                    self.tokenizer
+                        .clone()
```

Comment:
nit - we could reuse the tokenizer if we wanted by wrapping all inputs in a stream

Reply:
`.token_stream()` takes a `&str`, so I don't think that would be possible without combing all strings into a Mega String™